### PR TITLE
[Merged by Bors] - Add field docs for mirrord-config

### PR DESCRIPTION
**Pull request by meowjesty**
- Adds documentation for the many configurations of mirrord;
- Part of #538

### CODE REVIEW

File: mirrord-config/src/agent.rs [hunk 1]
```diff
@@ -3,32 +3,52 @@ use serde::Deserialize;
...
     #[config(env = "MIRRORD_AGENT_RUST_LOG", default = "info")]
     pub log_level: Option<String>,
 
+    /// Namespace where the agent shall live.
     #[config(env = "MIRRORD_AGENT_NAMESPACE")]
     pub namespace: Option<String>,
 
+    /// Name of the agent's docker image.
+    ///
+    /// You probably don't need this, unless you want to use a custom mirrord-agent build.
```
Comment (infiniteregrets) [inline mirrord-config/src/agent.rs:24]: I think we should avoid using pronouns like "You", just keep it formal like, "Not needed, unless a custom mirrord-agent build is required" or something along the lines

File: mirrord-config/src/agent.rs [hunk 2]
```diff
@@ -3,32 +3,52 @@ use serde::Deserialize;
...
     pub ephemeral: Option<bool>,
 
+    /// Controls how long should the agent await with no communication before timing out.
+    ///
+    /// mirrord has its own heartbeat mechanism, so even if your application has no messages, the
+    /// agent will stay alive until a heartbeat message times out with the value specified here.
     #[config(env = "MIRRORD_AGENT_COMMUNICATION_TIMEOUT")]
     pub communication_timeout: Option<u16>,
 
+    /// How long to wait for the agent to finish initialization.
```
Comment (infiniteregrets) [inline mirrord-config/src/agent.rs:51]: probably keep it along the lines of what you mentioned in communication timeout, start with "Controls.."

File: mirrord-config/src/env.rs [hunk 3]
```diff
@@ -8,17 +8,45 @@ use crate::{
...
+/// include = "project;database"
+/// ```
+///
+/// - Exclude the remote pod's environment variables "project", "database", and include everything
+///   else:
+///
+/// ```yaml
+/// # mirrord-config.yaml
+///
+/// exclude = "project;database"
```
Comment (infiniteregrets) [inline mirrord-config/src/env.rs:33]: maybe add the "*" example here?

**Review by infiniteregrets — COMMENTED** [review]
Looking great! If possible, could you please add a short note on how someone would go about adding a new field to the config? 

On a side note, the README is also named wrong for this package, can you also please update that?

File: mirrord-config/src/agent.rs [hunk 4]
```diff
@@ -3,32 +3,53 @@ use serde::Deserialize;
...
+    ///
+    /// Supports anything that would work with `RUST_LOG`.
     #[config(env = "MIRRORD_AGENT_RUST_LOG", default = "info")]
     pub log_level: Option<String>,
 
+    /// Namespace where the agent shall live.
     #[config(env = "MIRRORD_AGENT_NAMESPACE")]
     pub namespace: Option<String>,
 
+    /// Name of the agent's docker image.
```
Comment (eyalb181) [inline mirrord-config/src/agent.rs:22]: Add that it defaults to the latest stable image

File: mirrord-config/src/agent.rs [hunk 5]
```diff
@@ -3,32 +3,53 @@ use serde::Deserialize;
...
+    ///
+    /// Useful when a custom build of mirrord-agent is required.
```
Comment (aviramha) [inline mirrord-config/src/agent.rs:24]: or when using internal registry

File: mirrord-config/src/agent.rs [hunk 6]
```diff
@@ -3,32 +3,53 @@ use serde::Deserialize;
...
     pub image: Option<String>,
 
+    /// Controls when a new agent image is downloaded.
+    ///
+    /// Supports any valid kubernetes [image pull
+    /// policy](https://kubernetes.io/docs/concepts/containers/images/#image-pull-policy)
     #[config(env = "MIRRORD_AGENT_IMAGE_PULL_POLICY", default = "IfNotPresent")]
     pub image_pull_policy: Option<String>,
 
+    /// Controls for how long should the agent pod persist (even after the local process finished).
```
Comment (eyalb181) [inline mirrord-config/src/agent.rs:35]:
```suggestion
    /// Controls how long the agent pod persists for after the local process terminated.
```

File: mirrord-config/src/agent.rs [hunk 7]
```diff
@@ -3,32 +3,53 @@ use serde::Deserialize;
...
 #[serde(deny_unknown_fields)]
 #[config(map_to = AgentConfig)]
 pub struct AgentFileConfig {
+    /// Log level for the agent.
+    ///
+    /// Supports anything that would work with `RUST_LOG`.
     #[config(env = "MIRRORD_AGENT_RUST_LOG", default = "info")]
     pub log_level: Option<String>,
 
+    /// Namespace where the agent shall live.
```
Comment (aviramha) [inline mirrord-config/src/agent.rs:18]: Defaults to current k8s namespace

File: mirrord-config/src/agent.rs (hunk 6 above)
Comment (aviramha) [inline mirrord-config/src/agent.rs:35]: Can be useful for collecting logs.

File: mirrord-config/src/agent.rs [hunk 8]
```diff
@@ -3,32 +3,53 @@ use serde::Deserialize;
...
+    /// Controls for how long should the agent pod persist (even after the local process finished).
     #[config(env = "MIRRORD_AGENT_TTL", default = "0")]
     pub ttl: Option<u16>,
 
+    /// Runs the agent as an [ephemeral
+    /// container](https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/)
     #[config(env = "MIRRORD_EPHEMERAL_CONTAINER", default = "false")]
     pub ephemeral: Option<bool>,
 
+    /// Controls how long should the agent wait when there is no communication, before timing out.
```
Comment (eyalb181) [inline mirrord-config/src/agent.rs:44]:
```suggestion
    /// Controls how long the agent should wait when there is no communication, before timing out.
```

File: mirrord-config/src/agent.rs [hunk 9]
```diff
@@ -3,32 +3,53 @@ use serde::Deserialize;
...
+    ///
+    /// mirrord has its own heartbeat mechanism, so even if the local application has no messages,
+    /// the agent will stay alive until a heartbeat message times out with the value specified
```
Comment (aviramha) [inline mirrord-config/src/agent.rs:47]: not quite correct; this is how much time the agent will live without any connection, and each connection has a heartbeat.

File: mirrord-config/src/env.rs [hunk 10]
```diff
@@ -8,17 +8,56 @@ use crate::{
...
+    /// Include only these remote environment variables in the local process.
     #[config(env = "MIRRORD_OVERRIDE_ENV_VARS_INCLUDE")]
     pub include: Option<VecOrSingle<String>>,
 
+    /// Include the remote environment variables in the local process that are **NOT** specified by
+    /// this option.
     #[config(env = "MIRRORD_OVERRIDE_ENV_VARS_EXCLUDE")]
     pub exclude: Option<VecOrSingle<String>>,
 
+    // TODO(alex) [mid] 2022-10-29: Have no idea about this.
```
Comment (aviramha) [inline mirrord-config/src/env.rs:60]: This is a replacement for `.env` file or setting your own env then adding exclude.
For example, you want to set `AWS_REGION=1` but the remote pod has `AWS_REGION=2`
To have `AWS_REGION=1` applied, you need to set it at the local process + set it to exclude.
This is a shortcut to do it in one spot.

File: mirrord-config/src/agent.rs [hunk 11]
```diff
@@ -3,32 +3,53 @@ use serde::Deserialize;
 
 use crate::config::source::MirrordConfigSource;
 
+/// Configuration for the agent pod that is spawned in the kubernetes context.
```
Comment (eyalb181) [inline mirrord-config/src/agent.rs:6]:
```suggestion
/// Configuration for the mirrord-agent pod that is spawned in the Kubernetes cluster.
```

File: mirrord-config/src/lib.rs [hunk 12]
```diff
@@ -22,39 +22,52 @@ use crate::{
...
+    /// Allows mirrord to skip unwanted processes.
+    ///
+    /// Useful when process A spawns process B, and the user wants mirrord to operate only on
+    /// process B.
     #[config(env = "MIRRORD_SKIP_PROCESSES")]
     pub skip_processes: Option<VecOrSingle<String>>,
 
+    /// The target that will be impersonated by mirrord.
+    ///
+    /// Supports `pod`, `podname`, `deployment`, `container`, `containername`.
```
Comment (aviramha) [inline mirrord-config/src/lib.rs:47]: not entirely true, you can't just use `container/name`. This needs to be better explained

File: mirrord-config/src/lib.rs [hunk 13]
```diff
@@ -22,39 +22,52 @@ use crate::{
...
     #[config(env = "MIRRORD_IMPERSONATED_TARGET")]
     pub target: Option<String>,
 
+    /// Namespace where the `target` lives.
```
Comment (aviramha) [inline mirrord-config/src/lib.rs:51]: Defaults to current k8s config namespace

File: mirrord-config/src/env.rs [hunk 14]
```diff
@@ -8,17 +8,56 @@ use crate::{
     util::{MirrordToggleableConfig, VecOrSingle},
 };
 
+/// Allows the user to set or override a local process' environment variables with the ones from a
+/// remote pod.
+///
+/// Which environment variables to load from the remote are controlled by setting either `include`
```
Comment (eyalb181) [inline mirrord-config/src/env.rs:14]:
```suggestion
/// Which environment variables to load from the remote pod are controlled by setting either `include`
```

File: mirrord-config/src/network.rs [hunk 15]
```diff
@@ -10,18 +10,25 @@ use crate::{
...
+/// Controls mirrord network operations.
 #[derive(MirrordConfig, Deserialize, Default, PartialEq, Eq, Clone, Debug)]
 #[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
 #[serde(deny_unknown_fields)]
 #[config(map_to = NetworkConfig)]
 pub struct NetworkFileConfig {
+    /// Mode of operation for incoming network requests in mirrord, supports `mirror` or `steal`:
+    ///
+    /// - `mirror`: mirror requests from a Kubernetes container to the local process;
+    /// ` `steal`: redirects requests from the local process;
```
Comment (aviramha) [inline mirrord-config/src/network.rs:22]: to the local process*

File: mirrord-config/src/env.rs [hunk 16]
```diff
@@ -8,17 +8,56 @@ use crate::{
...
+/// or `exclude`.
+///
+/// ## Examples
+///
+/// - Include every environment variable from the remote pod (default).:
```
Comment (eyalb181) [inline mirrord-config/src/env.rs:19]:
```suggestion
/// - Include every environment variable from the remote pod (default):
```

**Review by aviramha — CHANGES_REQUESTED** [review]
Great job, see comments inline.

File: mirrord-config/src/env.rs [hunk 17]
```diff
@@ -8,17 +8,56 @@ use crate::{
...
+/// ```yaml
+/// # mirrord-config.yaml
+///
+/// include = "*"
+/// ```
+///
+/// Some environment variables are excluded by default (`PATH` for example), including these
+/// requires specifying them with `include`, see [`mirrord_agent::env::EnvFilter`].
+///
+/// - Include the remote pod's environment variables "project", "database":
```
Comment (eyalb181) [inline mirrord-config/src/env.rs:30]: Perhaps style them PROJECT, DATABASE?

File: mirrord-config/src/env.rs [hunk 18]
```diff
@@ -8,17 +8,56 @@ use crate::{
...
+/// # mirrord-config.yaml
+///
+/// exclude = "project;database"
+/// ```
 #[derive(MirrordConfig, Default, Deserialize, PartialEq, Eq, Clone, Debug)]
 #[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
 #[serde(deny_unknown_fields)]
 #[config(map_to = EnvConfig)]
 pub struct EnvFileConfig {
+    /// Include only these remote environment variables in the local process.
```
Comment (eyalb181) [inline mirrord-config/src/env.rs:51]:
```suggestion
    /// Include only these remote environment variables in the local process. Value is a list separated by ;.
```

File: mirrord-config/src/fs/advanced.rs [hunk 19]
```diff
@@ -16,18 +16,51 @@ use crate::{
...
+    /// The regexes specified here will make mirrord operate only on files that match it, otherwise
+    /// the file will be operated locally (bypassing mirrord).
     #[config(env = "MIRRORD_FILE_FILTER_INCLUDE")]
     pub include: Option<VecOrSingle<String>>,
 
+    /// Allows the user to specify regexes that are used to match against files when mirrord file
+    /// operations are enabled.
+    ///
+    /// The opposite of `include`, files that match the regexes specified here will bypass mirrord
+    /// and are operated locally.
```
Comment (eyalb181) [inline mirrord-config/src/fs/advanced.rs:63]:
```suggestion
    /// and are accessed locally.
```

File: mirrord-config/src/network.rs (hunk 15 above)
Comment (eyalb181) [reply mirrord-config/src/network.rs:22]: Should be :
- `mirror`: mirror incoming requests to the remote pod to the local process
- `steal`: redirect incoming requests to the remote pod to the local process

File: mirrord-config/src/network.rs [hunk 20]
```diff
@@ -10,18 +10,25 @@ use crate::{
...
     #[config(env = "MIRRORD_AGENT_TCP_STEAL_TRAFFIC", default = "mirror")]
     pub incoming: Option<IncomingConfig>,
 
+    /// Enables mirrord to take control of outgoing network operations.
```
Comment (eyalb181) [inline mirrord-config/src/network.rs:26]:
```suggestion
    /// Tunnel outgoing network operations through mirrord.
```

File: mirrord-config/src/network.rs [hunk 21]
```diff
@@ -10,18 +10,25 @@ use crate::{
...
     #[serde(default)]
     #[config(nested)]
     pub outgoing: ToggleableConfig<OutgoingFileConfig>,
 
+    /// Enables mirrord to resolve DNS via the remote pod.
```
Comment (eyalb181) [inline mirrord-config/src/network.rs:31]:
```suggestion
    /// Resolve DNS via the remote pod.
```

File: mirrord-config/src/feature.rs [hunk 22]
```diff
@@ -6,19 +6,23 @@ use crate::{
     network::NetworkFileConfig, util::ToggleableConfig,
 };
 
+/// Configuration for mirrord features.
 #[derive(MirrordConfig, Deserialize, Default, PartialEq, Eq, Clone, Debug)]
 #[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
 #[serde(deny_unknown_fields)]
 #[config(map_to = FeatureConfig)]
 pub struct FeatureFileConfig {
+    /// Controls the environment variables feature, see [`EnvFileConfig`].
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord-config/src/feature.rs:15]: I was following the naming convention, but ended with some `FileFilterFileConfig`, which was a very confusing name imo. Should I change it back? Should I change everything to be `XUserConfig`? Any other name (suggestions)?

[the remark this answers is not in the document]
Comment (eyalb181) [reply mirrord-config/src/feature.rs:15]: Lets stick with the FileConfig convention across the board (please do it in a separate PR though)

File: mirrord-config/src/lib.rs (hunk 12 above)
Comment (aviramha) [reply mirrord-config/src/lib.rs]: it doesn't support providing just a container
It's
`pod/{sample-pod}/[container]/container`
`deployment/{sample-deployment}/[container]/container`

Comment (meowjesty) [reply mirrord-config/src/lib.rs]:
```rust
/// Supports:
/// - `pod/{sample-pod}/[container]/{sample-container}`;
/// - `podname/{sample-pod}/[container]/{sample-container}`;
/// - `deployment/{sample-deployment}/[container]/{sample-container}`;
```

like so?

Comment (aviramha) [reply mirrord-config/src/lib.rs]: yup

**Review by eyalb181 — APPROVED** [review]
Incredible

**Comment by meowjesty** [comment]
Hey bors, I have a task for you (do it on your time, no rush)!

bors r+

**Review by aviramha — APPROVED** [review]

**Review by infiniteregrets — APPROVED** [review]

**Comment by meowjesty** [comment]
Sorry bors, I didn't mean to hurt you!

bors r+

**Pull request closed without merging**